csrc/ops/composite.cpp

```diff
@@ -54,42 +54,66 @@ TensorView* dropout_backward(TensorView* dy, TensorView* mask, Val* scale) {
-      "Only 2-D Inputs and Weights are currently supported in Linear!");
-
-  std::vector<bool> bcast_dims(a->nDims() + 1, false);
-  // A: [M, Bcast, K]
-  // B: [Bcast, N, K]
-  bcast_dims.at(bcast_dims.size() - 2) = true;
-  auto* tv0b = broadcast(a, bcast_dims);
-  bcast_dims.at(bcast_dims.size() - 2) = false;
-  bcast_dims.at(bcast_dims.size() - 3) = true;
-  auto* tv1b = broadcast(b, bcast_dims);
+namespace {
 
-  NVF_CHECK(
-      a->getDataType().value() == b->getDataType().value(),
-      "data types of inputs to matmul don't match");
-
-  auto* output = fusedMultiplySum(tv0b, tv1b, {-1});
-  if (bias) {
-    NVF_CHECK(
-        (bias->nDims() <= a->nDims()), "bias should be broadcastable to A");
-    NVF_CHECK(
-        a->getDataType().value() == bias->getDataType().value(),
-        "bias doesn't match input/weight dtype");
-    auto* bias_with_cast = maybeCastOp(output->getDataType().value(), bias);
-    auto* bcast_bias = ops::maybeBroadcast({output, bias_with_cast})[1];
-    auto* bias_output = add(output, bcast_bias);
-    return maybeCastOp(a->getDataType().value(), bias_output);
+static TensorView* newForLinear(
+    TensorView* input,
+    TensorView* weight,
+    TensorView* bias) {
+  auto input_domain =
+      TensorDomain::noReductions(input->getMaybeRFactorDomain());
+  auto weight_domain =
+      TensorDomain::noReductions(weight->getMaybeRFactorDomain());
+
+  // Linear: inputs = {*, in_features}, weight = {out_features, in_features} /
+  // {in_features} For the linear output, all but the last dimension are the
+  // same shape as the input. The last dimension is out_features (if present).
+  auto ndims_out = (input_domain.size() - 1) + (weight_domain.size() - 1);
+
+  std::vector<IterDomain*> out_domain(ndims_out, nullptr);
+
+  for (auto idx : c10::irange(input_domain.size() - 1)) {
+    out_domain[idx] = ops::newOutputIterDomain({input_domain.at(idx)});
   }
-  return maybeCastOp(a->getDataType().value(), output);
+
+  if (weight_domain.size() == 2) {
+    // Add out_features to output domain.
+    std::vector<IterDomain*> out_features_ids = {weight_domain.at(0)};
+    if (bias != nullptr) {
+      auto bias_domain =
+          TensorDomain::noReductions(bias->getMaybeRFactorDomain());
+      out_features_ids.emplace_back(bias_domain.at(0));
+    }
+    out_domain[ndims_out - 1] = ops::newOutputIterDomain(out_features_ids);
+  }
+}
+
+TensorDomain* td = IrBuilder::create<TensorDomain>(
+    out_domain,
+    TensorDomain::getContiguityFilledWith(out_domain, true));
+
+return IrBuilder::create<TensorView>(td, input->dtype());
+}
+
+} // namespace
+
+TensorView* linear(TensorView* tv_a, TensorView* tv_b, TensorView* bias) {
+  NVF_CHECK(tv_a->nDims() >= 1, "Input A must be atleast 1D.");
+  NVF_CHECK(
+      tv_b->nDims() == 1 || tv_b->nDims() == 2,
+      "Input B must be a 1D / 2D tensor.");
+  NVF_CHECK(
+      tv_b->nDims() == 1 && bias != nullptr,
+      "Input B must be a 2D tensor if bias is present, got 1D.")
```

Comment:
I think this condition should be `bias == nullptr || (tv_b->nDims() == bias->nDims() + 1)`. Also note though that `nDims()` is the size of the leaf domain which includes reductions. In an unsegmented fusion we might actually have reductions that feed straight to matmul or linear, so the dims should be computed via `TensorDomain::noReductions(tv->getMaybeRFactorDomain()).size()` instead.

Reply:
```
>>> a = torch.randn(2, 2, dtype=torch.float16)
>>> b = torch.randn(2, dtype=torch.float16)
>>> bias = torch.tensor(3.14, dtype=torch.float16)
>>> torch.nn.functional.linear(a, b, bias)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: mat2 must be a matrix, got 1-D tensor
```

Bias is only accepted when B (weight) is a matrix.

Reply:
Thanks for checking! That constraint is not mentioned in the docs. Let's put that in a comment that although the docs don't mention it, as of now bias is restricted to the case of 2D weight arguments.

Reply:
Another case that is supported is a 2D bias although the documents do not mention it (The current mapping will not support this case): `A = [2, 4, 6], B [3, 6], Bias = [2*4, 3] `.

Reply:
Should the bias not be [ 4, 3 ] in that case?

Reply:
No, that fails actually. It looks for the merge of all the non-reduction dimensions in the first operand.

Reply:
The crazy thing is that this is particular to 3D input. If you have 4D input then you can use a 4D bias without any merged dim as expected, but for some reason 3D input requires 2D bias where the first two dims are merged. This is something we could support if we want to, but since it's undocumented we could also just disallow 2D bias as we have here and prevent thunder from sending us such linears. wdyt @jjsjann123

Reply:
Thunder does not accept 2D bias: https://github.com/Lightning-AI/lightning-thunder/blob/2fa97d6b1fead1eb9d0dc17703ebb12bbf752bbc/thunder/core/prims.py#L3456